### PR TITLE
d&i MorphismsOfExternalHom

### PR DESCRIPTION
**Pull request by mohamed-barakat**
outsourced from Toposes

### CODE REVIEW

**Comment by zickgraf** [comment]
Could you increase the code coverage, maybe using the terminal category?

**Comment by mohamed-barakat** [comment]
> Could you increase the code coverage, maybe using the terminal category?

Done.

**Comment by mohamed-barakat** [comment]
Had to force push again to trigger the tests.

**Comment by mohamed-barakat** [comment]
At some point, we could think about making `TerminalCategoryWithSingleObject( )` the standard `RangeCategoryOfHomomorphismStructure` for all terminal categories and `AddMorphismsOfExternalHom` only there. This would trigger the uncovered derivation.

**Comment by zickgraf** [comment]
> > Could you increase the code coverage, maybe using the terminal category?
> 
> Done.

It seems like the coverage did not really increase.

**Comment by mohamed-barakat** [comment]
> It seems like the coverage did not really increase.

The coverage would increase if the derivation is triggered but I don't see a simple way to trigger it. It is of course triggered in `CategoricalTowers`.

**Comment by mohamed-barakat** [comment]
> At some point, we could think about making `TerminalCategoryWithSingleObject( )` the standard `RangeCategoryOfHomomorphismStructure` for all terminal categories and `AddMorphismsOfExternalHom` only there. This would trigger the uncovered derivation.

I realized this idea in multiple commits. If you like it I can split the PR if you wish.

**Comment by mohamed-barakat** [comment]
The update PR should include all your suggestions, if I am not forgetting something.

File: CAP/gap/CategoryConstructor.gi [hunk 1]
```diff
@@ -41,6 +41,7 @@ InstallMethod( CategoryConstructor,
         create_func_morphism := IsObject,
         create_func_morphism_or_fail := IsObject,
         create_func_list_of_objects := IsObject,
+        range_category_of_homomorphism_structure := cat -> IsCapCategory( cat ) or cat = "self",
```
Comment (zickgraf) [inline CAP/gap/CategoryConstructor.gi:44]: Please move this after `commutative_ring_of_linear_category`.

**Comment by zickgraf** [comment]
@mohamed-barakat This change breaks the CAP tests in Julia, could you look into this?

```
2024-01-03T04:27:38.5739965Z │ Subexpression:
2024-01-03T04:27:38.5740199Z │ 
2024-01-03T04:27:38.5740472Z │ Hom_ab = MorphismsOfExternalHom( a, b )
2024-01-03T04:27:38.5740796Z │ 
2024-01-03T04:27:38.5741015Z │ Evaluated output:
2024-01-03T04:27:38.5741255Z │ 
2024-01-03T04:27:38.5742216Z │ 1-element Vector{TheJuliaConcreteTypeIsMorphismInCapTerminalCategoryWithMultipleObjects{Symbol("TerminalCategoryWithMultipleObjects( )InstanceMorphismFilter##236")}}:
2024-01-03T04:27:38.5743309Z │  <A morphism in TerminalCategoryWithMultipleObjects( )>
2024-01-03T04:27:38.5743716Z │ 
2024-01-03T04:27:38.5743927Z │ Expected output:
2024-01-03T04:27:38.5744165Z │ 
2024-01-03T04:27:38.5744558Z │ [ <A zero, isomorphism in TerminalCategoryWithMultipleObjects( )> ]
2024-01-03T04:27:38.5745001Z │ 
```

https://github.com/homalg-project/FinSetsForCAP/actions/runs/7393434313/job/20113357003

**Pull request merged**